### PR TITLE
Add futures to ProcessGroupMPI work (but not including Send/Recv) and python DDP comm hook testing

### LINKED ISSUE #42048: [Gloo, MPI] DDP communication hook: getFuture()
DDP communication hook implementation #39272 provides an API called `get_future` to retrieve a future associated with the completion of `c10d.ProcessGroupNCCL.work`. `get_future` API is currently only supported by NCCL #41596 , because most of the intensive models, where our gradient compression effort might be more useful, use NCCL. 

To support Gloo and MPI, we can modify some of the core ProcessGroupGloo and ProcessGroupMPI code to ensure we can do something similar there.

cc @ezyang @gchanan @zou3519 @bdhirsh @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @xush6528

### PR DESCRIPTION
**Pull request by agolynski**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57214 Add futures to ProcessGroupMPI work (but not including Send/Recv) and python DDP comm hook testing**

Differential Revision: [D28200791](https://our.internmc.facebook.com/intern/diff/D28200791)

### CODE REVIEW

File: torch/testing/_internal/distributed/distributed_test.py
Comment (mrshenli) [inline torch/testing/_internal/distributed/distributed_test.py:3397]: Looks like this function is copied from `test_c10d_nccl.py`? If yes, let's dedup.

**Review by mrshenli — CHANGES_REQUESTED** [review]
Usually, the test needs to be in the same PR as the feature. Otherwise, we won't be confident if the feature is implemented correctly.

File: torch/testing/_internal/distributed/distributed_test.py
Comment (agolynski) [reply torch/testing/_internal/distributed/distributed_test.py:3426]: migrated comm hook NCCL tests from test_c10d_nccl.py to here

File: torch/testing/_internal/common_distributed.py
Comment (mrshenli) [inline torch/testing/_internal/common_distributed.py:108]: This means this decorator won't skip the test even if the ci env does not contain x or more GPUs?

File: torch/testing/_internal/distributed/distributed_test.py
Comment (mrshenli) [inline torch/testing/_internal/distributed/distributed_test.py:416]: why deleting one space here?

File: torch/testing/_internal/common_distributed.py
Comment (agolynski) [reply torch/testing/_internal/common_distributed.py:108]: correct, if only skips test if backend is nccl and we have < x GPUs

File: torch/testing/_internal/distributed/distributed_test.py
Comment (mrshenli) [inline torch/testing/_internal/distributed/distributed_test.py:3441]: Does this mean we only test MPI for CPU modules?

File: torch/testing/_internal/distributed/distributed_test.py
Comment (agolynski) [reply torch/testing/_internal/distributed/distributed_test.py]: fixed

File: torch/testing/_internal/distributed/distributed_test.py
Comment (agolynski) [reply torch/testing/_internal/distributed/distributed_test.py:3441]: filed https://github.com/pytorch/pytorch/issues/57547

File: torch/testing/_internal/common_distributed.py
Comment (mrshenli) [reply torch/testing/_internal/common_distributed.py:108]: I see. Is it possible to avoid introducing this additional decorator by having two test entries? E.g.:

```python
def _test_xyz(self, ...):
  ...

@unittest.skipIf(BACKEND == "nccl")
def test_xyz(self):
  self._test_xyz()

@unittest.skipIf(BACKEND != "nccl")
@skip_if_lt_x_gpu(2)
def test_xyz_nccl:
  self._test_xyz()

```

File: torch/lib/c10d/ProcessGroupMPI.cpp
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupMPI.cpp:110]: nit: curious about the naming. Why do we need both finish and complete in the function name? Do you mean `finishCompletedFuture`, or will `markCompleted` work?

File: torch/lib/c10d/ProcessGroupMPI.hpp
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupMPI.hpp:95]: could you please run clang-format on the new code? I think there needs to be two additional spaces here

File: torch/lib/c10d/ProcessGroupMPI.cpp
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupMPI.cpp:115]: Do we plan to cover send/recv APIs as well in this release, or do we plan to focus on collectives for now? If it's the later, let's create an issue to track.

**Review by mrshenli — COMMENTED** [review]
In general LGTM! Thanks for covering the gaps. I left some minor comments. Will stamp when all tests pass.

**Comment by mrshenli** [comment]
Could you please import this PR? I will stamp both

File: torch/lib/c10d/ProcessGroupMPI.cpp
Comment (agolynski) [reply torch/lib/c10d/ProcessGroupMPI.cpp:115]: let's focus on collectives. send/recv work will require changing how AsyncWork is handled. although we have a good idea what we'd like it to be (given @lw comments), it could be a breaking change for users who might rely on async/wait behavior.

will file an issue.

**Review by mrshenli — APPROVED** [review]
stamp to unblock. Please address comments of the last round of review before landing.

**Comment by mrshenli** [comment]
internal diff is D28162480

**Comment by agolynski** [comment]
> I see. Is it possible to avoid introducing this additional decorator by having two test entries? E.g.:

If we add more testing functions, I think it's more readable with decorator rather than duplicating them and adding more helpers?

File: torch/lib/c10d/ProcessGroupMPI.cpp
Comment (agolynski) [reply torch/lib/c10d/ProcessGroupMPI.cpp:110]: I meant that it does both, let me just call it finishWorkMPI() so that it's finishes whatever belongs to WorkMPI.

File: torch/testing/_internal/common_distributed.py
Comment (agolynski) [reply torch/testing/_internal/common_distributed.py:108]: If we add more testing functions, I think it's more readable with decorator rather than duplicating them and adding more helpers?

File: torch/lib/c10d/ProcessGroupMPI.hpp
Comment (agolynski) [reply torch/lib/c10d/ProcessGroupMPI.hpp:95]: Done. my clang ext in VS broke for some reason, sorry about that

File: torch/testing/_internal/common_distributed.py
Comment (mrshenli) [reply torch/testing/_internal/common_distributed.py:108]: It's up to you. :)

**Pull request closed without merging**
